### PR TITLE
feat(ngMock) add atomic flushing strategy for $timeout

### PR DESCRIPTION
**Pull request by twlvl**
Hi dear Angular community,

During writing unit tests I found one interesting issue. I need to write tests for some code like this:

``` javascript
$timeout(function() {
 action1();
 $timeout(function() {
  action2();
 });
});
```

And I'm expecting that I need to execute `$timeout.flush()`  2 times. First time to reach `action1` function and the second time to execute `action2` function (let's call this kind of flush _"atomic"_ for example). But for now, for the first flush() it executes the both handlers. One after another.  Please check this example: http://jsbin.com/fuziri/edit?js,output .

On the other hand we have some cases in tests when this buggy, for the first look, behaviour can simplify our lifes. In such case for example:

``` javascript
 var callback = jasmine.createSpy();
 $timeout(function() {}, 0, false).then(callback);
 $timeout.flush();
 expect(callback).toHaveBeenCalled();
```

Here `$timeout` register new defered task to process callback using `$browser.defer` during flushing. So to execute callback (if you are using that _atomic_ flush strategy) you should call `$timeout.flush()` twise.

I think that the both strategies of flushing are pretty useful. And may be it will be a good idea to:
- implement that _atomic_ flushing
- to allow to switch between the strategies using some additional flag

What do you think, guys?

P.S. this is a duplicate of #13146, I opened a new one because can't solve the certificate issue for commit with wrong email address.

### CODE REVIEW

File: src/ngMock/angular-mocks.js [hunk 1]
```diff
@@ -103,8 +103,11 @@ angular.mock.$Browser = function() {
    * Flushes all pending requests and executes the defer callbacks.
    *
    * @param {number=} number of milliseconds to flush. See {@link #defer.now}
+   * @param {boolean=} [atomic=false] If set to `true` — will handle only tasks
+   * available at the moment of execution, otherwise will process tasks that was
```
Comment (atogata) [inline src/ngMock/angular-mocks.js]: that were*

File: src/ngMock/angular-mocks.js [hunk 2]
```diff
@@ -1948,9 +1968,11 @@ angular.mock.$TimeoutDecorator = ['$delegate', '$browser', function($delegate, $
    * Flushes the queue of pending tasks.
    *
    * @param {number=} delay maximum timeout amount to flush up until
+   * @param {boolean=} [atomic=false] If set to `true` — will not run the tasks
+   * added during execution of flash even if they added with `0` timeout.
```
Comment (atogata) [inline src/ngMock/angular-mocks.js]: flush*
they were*

**Comment by atogata** [comment]
+1

**Comment by elkeis** [comment]
@atogata, thanks. Spelling corrected.

**Comment by petebacondarwin** [comment]
@elkeis - did you know that you can pass a `delay` to `$timeout.flush(delay)`, which will only flush timeouts that should have completed by the given delay. This is not quite what you want, perhaps, since if you have two timeouts with 0 delay then they would both always be flushed at the same time.
But this can help with testing sequencing issues.

**Comment by petebacondarwin** [comment]
Oh I just looked at the code. I see that you want to ensure that new calls to timeout that are triggered as part of flushing the current timeouts are not triggered...

**Comment by petebacondarwin** [comment]
Interesting idea. I think this would need some more API thought as the current `atomic` parameter is not that intuitive to me.

**Comment by petebacondarwin** [comment]
This is very vaguely related to #5420. It would be interesting to solve both at once...

**Comment by petebacondarwin** [comment]
@elkeis once #14686 lands then as long as your timeouts have a non-zero delay, then they will indeed become separable in tests.

Given your original example, if I add 1ms second delays:

```js
$timeout(function() {
 action1();
 $timeout(function() {
  action2();
 },1);
}, 1);
```

Then calling `$timeout.flush(1)` will only trigger `action1()`. Calling `$timeout.flush(1)` a second time will trigger `action2()`.

Is this enough for your testing?

**Comment by petebacondarwin** [comment]
Closing as #14686 has landed and there has been activity on this PR for 10 months.

**Pull request closed without merging**